doc/developer/design/20250203_cluster_reliability.md

```diff
@@ -0,0 +1,224 @@
+
+- Associated: MaterializeInc/materialize#31246 (MVP)
+
+<!--
+The goal of a design document is to thoroughly discover problems and
+examine potential solutions before moving into the delivery phase of
+a project. In order to be ready to share, a design document must address
+the questions in each of the following sections. Any additional content
+is at the discretion of the author.
+
+Note: Feel free to add or remove sections as needed. However, most design
+docs should at least keep the suggested sections.
+-->
+
+## The Problem
+
+<!--
+What is the user problem we want to solve?
+
+The answer to this question should link to at least one open GitHub
+issue describing the problem.
+-->
+
+Materialize offers little mechanisms to ensure a cluster stays up once it is running, which is at odds with the requirements for operational systems.
+Users expect an operational system to be always available, and giving the mechanisms to prevent users from compromising the system.
+The underlying issue is that Materialize does not offer performance isolation within a replica.
+Any user with the ability to create dataflows, or read from external storage, can create objects that consume the remaining resources and cause unavailability.
+
+Materialize offers coarse-grained mechanisms to prevent some of the issues.
+RBAC can express a policy that prevents users from issuing queries, or creating expensive objects.
+It does not, however, allow fine-grained permissions: A query that reads from an index causes much less work than a query that requires its own dataflow.
+In this design, we look at a spectrum of solutions to address this issue.
+
+
+## Success Criteria
+
+<!--
+What does a solution to this problem need to accomplish in order to
+be successful?
+
+The criteria should help us verify that a proposed solution would solve
+our problem without naming a specific solution. Instead, focus on the
+outcomes we hope result from this work. Feel free to list both qualitative
+and quantitative measurements.
+-->
+
+Materialize offers a range of mechanisms that allow users to prevent undesired operations on a cluster.
+
+## Out of Scope
+
+<!--
+What does a solution to this problem not need to address in order to be
+successful?
+
+It's important to be clear about what parts of a problem we won't be solving
+and why. This leads to crisper designs, and it aids in focusing the reviewer.
+-->
+
+We distinguish transient from permanent dataflows for the purposes of this design.
+While we could apply the same mechanism to permanent dataflows, it is less clear how we would operate the system in the presence of a limit violation.
+We cannot simply terminate a permanent dataflow, as it would affect all queries that depend on it, and our users expect dataflows to be always-on.
+For this reason, we limit the scope of this design to transient dataflows, which are selects and subscribes.
+
+We also do not address the problem how the user can express the limit.
```

Comment:
While RBAC doesn't talk about limits, a given user could have quotas (just like UNIX limits/quotas). Would allow them to say that `dbt` gets to use arbitrary memory for its ad hoc work, but someone on the data team has limits about how they interact with the production cluster/replicas.